content/1_docs/2_cookbook/3_templating/0_understanding-oop/recipe.txt

```diff
@@ -0,0 +1,462 @@
+Title: A brief intro to object oriented programming in PHP
+
+----
+
+Published: 2021-01-11
+
+----
+
+Description: Understanding classes, objects, member functions and common pitfalls
+
+----
+
+Text:
+
+## Intro
+
+When you have already tinkered with Kirby for a while, you are likely to have come across terms like classes, objects or methods. You've probably run into errors like "calling a member function x on null" and have then been asked to "check for an object before you call a method". Or have tried to use `$kirby` or `$site` in a plugin and been told to use `kirby()` and `site()` in this context instead. And then you've followed the advice but never really understood why.
+
+If that is true, then this recipe is for you. We will look into the basics of object oriented programming (OOP) with the goal of making you understand what this is all about.
+
+The Kirby core is written in PHP, a scripting language that still powers a majority of websites both in the backend and in the frontend.
+
+Like basically all professional PHP applications today, Kirby is based on an OOP (Object Oriented Programming) architecture which has many advantages over the procedural approach you often find in older PHP programs. The idea behind this architecture is to encapsulate data and the methods that work with this data, in order to produce modular, extendable and maintainable code.
+
+<warning>
+No images, lots of text😉. But at least some code to play with. So get some coffee or tea.
+</warning>
+
+## Prerequisites
+
+If you want to code along with this recipe, you need the following:
+
+- A web server (ideally on your (link: docs/cookbook/setup/development-environment text: local machine))
+- A current version of PHP
+- A Kirby installation , either (link: docs/guide/quickstart text: Starterkit or Plainkit)
+- A code/text editor of your choice
+
+## What is OOP?
+
+Object Oriented Programming (OOP) is all about objects, their properties and their methods.
+
+Let's consider a real-world object. Such an object could be a book in your library (or a computer, a smartphone, a piece of furniture…). It has certain properties like an author, a title, an ISBN number, a price tag and a cover, and it was probably published by a publishing house. Its methods would be the actions you could do with the book, e.g. you can read it, you can give it away as a present, put it back in your shelf or throw it away if you don't like it. And while all the books in your library are different, they nevertheless share the same general properties and methods.
```

Comment:
I really like the more detailed examples for the book object! I'd leave out the other objects though as they are confusing before the book explanations. Alternative: Put the other objects at the end like this:

```suggestion
Let's consider a real-world object. Such an object could be a book in your library. It has certain properties like an author, a title, an ISBN number, a price tag and a cover, and it was probably published by a publishing house. Its methods would be the actions you could do with the book, e.g. you can read it, you can give it away as a present, put it back in your shelf or throw it away if you don't like it. And while all the books in your library are different, they nevertheless share the same general properties and methods. Other objects could be a computer, a smartphone or a piece of furniture. All have their own properties and methods.
```